### DISCUSSION TITLE
How to run Semantic Kernel in Production

### DISCUSSION THREAD

**Discussion by timlaverty** [microsoft/semantic-kernel#136] [Q&A]
@dluc @craigomatic 

Lots of discussion about this amongst the team so far but we'd like community input on it.  My mental model is folks would run SK inside an Azure service (either ACI, AKS, or other) and put keys in keyvault.  The client app (typescript, wpf, etc.) would talk to the cloud service when making its AI calls.  If everything has to be in a thick client app, I still picture pulling keys from a cloud vault and never storing them on client.

I think we need a write a paper on, “How to run Semantic Kernel in Production” in our repo.  I'd love input from folks as they build apps and deploy them - that practical experience would be good influence.

-Tim

**Comment by dluc** [comment]
While we brainstorm the topic and find the time to share it, here are some high level notes:

* Bootstrapping a kernel instance is cheap, and we assume that the typical use case will be creating a new instance whenever there's a need to: create a plan, execute a plan, access memory, create and execute a semantic function on the fly. Even during the execution of a plan, is reasonable to create a new instance step by step.
* The kernel is thread safe, so it can support the execution of multiple plans in parallel and isolation. This should be properly verified, but that's the underlying intention. 
* The kernel does not support concurrent users though. The kernel state (e.g. the Memory instance), currently, is user specific. This can be addressed with custom memory implementations, or limiting the memory interaction to native skills, but it's not recommended, and error-prone. For user specific Memory, we should expand the context object first, e.g. allow contextual memory. This is not in the current roadmap, hence the recommendation to not use singleton kernel instances, and to discard kernel instances in multi-user environments/services.
* Native skills should be designed to be thread safe and support multiple users. The kernel cannot verify how skills behave though, and the current kernel holds skill instances active, for performance reasons. Instantiating skills on demand, e.g using a factory, could help, but this is not in the roadmap, and potentially not a responsibility of the kernel. SK Core skills are designed to support concurrent users, but third party skills might not, so the right approach depends on which skills are used and their internal design.

**Comment by timlaverty** [comment]
Running SK in Azure hosted by an Azure Function:  https://devblogs.microsoft.com/semantic-kernel/how-to-deploy-semantic-kernel-to-azure-in-minutes/.  Note this means any SK Skills you use need to be stateless.  All the Skills "inbox" are, but any you add would need to be stateless too.

An option for stateful hosting would be ASP.NET.

**Comment by AceHack** [comment]
Can you expand on what it means that the memory instance is user specifics?  What is the memory instance?  What is it used for?  Thanks.

**Comment by kinthaiofficial** [comment]
Running Semantic Kernel in production surfaces a few considerations that aren't obvious from the docs:

**Connection pooling and retry policies**
The default HttpClient lifecycle in .NET can cause port exhaustion at scale. Use `IHttpClientFactory` with Semantic Kernel's HTTP client registration rather than creating clients per-invocation. Also configure retry policies (Polly is integrated) for the inevitable transient API failures.

**Semantic function versioning**
Prompt templates evolve. When you update a semantic function, the behavior changes silently for callers who don't know. Consider versioning your skill directories: `skills/v1/SummarySkill/`, `skills/v2/SummarySkill/`, and routing calls to the appropriate version explicitly.

**Memory store selection for scale**
The in-memory memory store is fine for demos; for production you need Azure Cognitive Search, Pinecone, Qdrant, or similar. The key production requirement is persistence across restarts and vector search that scales beyond RAM.

**Planner timeout configuration**
The planner can make multiple LLM calls to construct a plan, and each call has latency. Set explicit timeouts and fall back to simpler execution paths when the planner exceeds them — otherwise a slow API day brings down your planning-heavy workflows.

**Observability hooks**
SK has event hooks you can attach to semantic function invocations. Wire these to your APM (Application Insights, Datadog) to get: function invocation counts, latency per function, token usage per invocation, error rates. Without this data you're flying blind on production cost and performance.

**Kernel instance scoping**
In ASP.NET, scope the Kernel to the request (not singleton) if your skills use request-specific context. Singleton Kernels accumulate state across requests in unexpected ways.

What's the scale you're targeting — requests/day?
